File: packages/nerva-py/nerva/tools/composite.py
"""Composite tool manager — combine multiple ToolManagers into one (N-613).

Merges discovery results from all inner managers (deduplicated by name)
and routes tool calls to the manager that owns each tool. First manager
to claim a tool name wins (priority order).
"""

from __future__ import annotations

import logging
import time
from typing import TYPE_CHECKING

from nerva.tools import ToolManager, ToolResult, ToolSpec, ToolStatus

if TYPE_CHECKING:
    from nerva.context import ExecContext

__all__ = ["CompositeToolManager"]

_log = logging.getLogger(__name__)


class CompositeToolManager:
    """Combine multiple ToolManagers into a single unified interface.

    Discovery merges results from all managers, deduplicating by tool name
    (first manager to claim a name wins). Tool calls are routed to the
    owning manager.

    Args:
        managers: Ordered list of ToolManagers. Priority is first-to-last.
    """

    def __init__(self, managers: list[ToolManager]) -> None:
        self._managers = list(managers)
        self._tool_owner: dict[str, ToolManager] = {}

    # -- Public API ----------------------------------------------------------

    async def discover(self, ctx: ExecContext) -> list[ToolSpec]:
        """Discover tools from all inner managers, deduplicated by name.

        First manager to register a tool name wins. Subsequent managers
        with the same tool name are silently ignored.

        Args:
            ctx: Execution context carrying identity and permissions.

        Returns:
            Merged, deduplicated list of tool specs.
        """
        seen_names: set[str] = set()
        merged: list[ToolSpec] = []

        for manager in self._managers:
            specs = await _safe_discover(manager, ctx)
            for spec in specs:
                if spec.name in seen_names:
                    continue
                seen_names.add(spec.name)
                merged.append(spec)
                self._tool_owner[spec.name] = manager

        return merged

    async def call(
        self, tool: str, args: dict[str, object], ctx: ExecContext
    ) -> ToolResult:
        """Route a tool call to the manager that owns it.

        Args:
            tool: Tool name to invoke.
            args: Arguments matching the tool's parameter schema.
            ctx: Execution context carrying identity and permissions.

        Returns:
            ToolResult from the owning manager, or NOT_FOUND if unowned.
        """
        owner = self._tool_owner.get(tool)
        if owner is None:
            return ToolResult(
                status=ToolStatus.NOT_FOUND,
                error=f"Tool '{tool}' not found in any manager",
            )

        return await owner.call(tool, args, ctx)


# -- Helpers -----------------------------------------------------------------


async def _safe_discover(manager: ToolManager, ctx: ExecContext) -> list[ToolSpec]:
    """Discover tools from a manager, suppressing errors.

    Args:
        manager: The ToolManager to query.
        ctx: Execution context.

    Returns:
        Tool specs from the manager, or empty list on failure.
    """
    try:
        return await manager.discover(ctx)
    except Exception as exc:  # noqa: BLE001
        _log.warning("Tool discovery failed for %s: %s", type(manager).__name__, exc)
        return []

File: packages/nerva-js/src/tools/composite.ts
/**
 * Composite tool manager — combines multiple ToolManagers into one.
 *
 * Merges tool discovery across managers and routes `call()` to the
 * owning manager. Priority: the first manager to register a tool name wins.
 *
 * @module tools/composite
 */

import type { ExecContext } from "../context.js";
import {
  ToolStatus,
  type ToolSpec,
  type ToolResult,
  type ToolManager,
  createToolResult,
} from "./index.js";

// ---------------------------------------------------------------------------
// CompositeToolManager
// ---------------------------------------------------------------------------

/**
 * Combines multiple ToolManagers into a single unified interface.
 *
 * On `discover()`, merges tools from all managers, deduplicating by name
 * (first manager to claim a name wins). On `call()`, routes to the
 * manager that owns the requested tool.
 *
 * @example
 * ```ts
 * const composite = new CompositeToolManager([mcpManager, functionManager]);
 * const tools = await composite.discover(ctx);
 * const result = await composite.call("calculator", { a: "1", b: "2" }, ctx);
 * ```
 */
export class CompositeToolManager implements ToolManager {
  private readonly _managers: readonly ToolManager[];
  private readonly _toolOwner = new Map<string, ToolManager>();

  /**
   * @param managers - Ordered list of tool managers. Earlier managers have higher priority.
   */
  constructor(managers: readonly ToolManager[]) {
    this._managers = managers;
  }

  /**
   * Discover tools from all managers, deduplicating by name.
   *
   * The first manager to provide a tool with a given name wins.
   * Subsequent managers with the same tool name are skipped.
   *
   * @param ctx - Execution context carrying identity and permission set.
   * @returns Merged and deduplicated list of tool specs.
   */
  async discover(ctx: ExecContext): Promise<ToolSpec[]> {
    const seen = new Set<string>();
    const result: ToolSpec[] = [];

    for (const manager of this._managers) {
      const tools = await manager.discover(ctx);
      for (const tool of tools) {
        if (seen.has(tool.name)) {
          continue;
        }
        seen.add(tool.name);
        result.push(tool);
        this._toolOwner.set(tool.name, manager);
      }
    }

    return result;
  }

  /**
   * Execute a tool call, routing to the owning manager.
   *
   * If the tool has not been seen during `discover()`, attempts to find
   * the tool by calling `discover()` first. Returns NOT_FOUND if no
   * manager owns the requested tool.
   *
   * @param tool - Tool name to invoke.
   * @param args - Arguments matching the tool's parameter schema.
   * @param ctx - Execution context carrying identity and permission set.
   * @returns ToolResult from the owning manager.
   */
  async call(
    tool: string,
    args: Record<string, unknown>,
    ctx: ExecContext,
  ): Promise<ToolResult> {
    let owner = this._toolOwner.get(tool);

    if (owner === undefined) {
      // Tool not yet discovered — try a fresh discovery pass
      await this.discover(ctx);
      owner = this._toolOwner.get(tool);
    }

    if (owner === undefined) {
      return createToolResult(ToolStatus.NOT_FOUND, {
        error: `Tool '${tool}' not found in any manager`,
      });
    }

    return owner.call(tool, args, ctx);
  }
}
